### ISSUE TITLE
msc placeholder: not afraid of assembly

### ISSUE THREAD

**Issue by synctext** [Tribler/tribler#7456]
_brainstorm placeholder_

Embedded systems master. Start Q2 or perhaps (possible) Literature Survey earlier.
On progress for Q1 finishing all course work+survey. Then Q2 start thesis full-time, ~nov.

Likes Rust, C coding. Completed machine learning, evolutionary algorithms (ongoing) {not deep learning}. Interested in decentralised machine learning. See our Delft/EPFL work: [MoDeST: Bridging the Gap between Federated and Decentralized Learning with Decentralized Sampling](https://arxiv.org/pdf/2302.13837.pdf). master thesis on arXiv: [G-Rank: Unsupervised Continuous Learn-to-Rank for Edge Devices in a P2P Network](https://arxiv.org/pdf/2301.12530.pdf). CODE: https://github.com/awrgold/G-Rank (Jupyter Notebook file)

Scholar Google search for anything related to "[LLM Android](https://scholar.google.com/scholar?q=LLM+android)"
[DroidBot-GPT: GPT-powered UI Automation for Android](https://arxiv.org/pdf/2304.07061). Clusion after 2 minutes of effort: nobody is combining LLM for Android yet. Homework for reader: "nanoGPT Android".
```
We evaluate DroidBot-GPT with a self-created dataset that contains 33 tasks collected
from 17 Android applications spanning 10 categories. It can successfully complete 39.39% of the
tasks, and the average partial completion progress is about 66.76%.
```

EDIT: another potential direction to explore, hardware-accelerated machine learning on Android https://towardsdatascience.com/gpu-accelerated-machine-learning-in-your-mobile-applications-using-the-android-ndk-vulkan-kompute-1e9da37b7617

EDIT2: lots of activity on federated learning and swarm learning. Enough for a "literature survey". Have a broader scope than thesis, gives a broad understanding, before doing a deep dive for thesis.
- https://deepai.org/publication/peer-to-peer-federated-learning-on-graphs
- https://deepai.org/publication/federated-learning-at-the-network-edge-when-not-all-nodes-are-created-equal
- https://deepai.org/publication/decentralized-federated-learning-a-segmented-gossip-approach
- https://deepai.org/publication/gfl-a-decentralized-federated-learning-framework-based-on-blockchain
- https://deepai.org/publication/a-blockchain-based-decentralized-federated-learning-framework-with-committee-consensus
- Sybil and security issues
  - "[Towards Sybil Resilience in Decentralized Learning](https://arxiv.org/pdf/2306.15044.pdf)" 
  - https://deepai.org/publication/collaborative-domain-blocking-using-federated-nlp-to-detect-malicious-domains
- Models
  - https://www.tensorflow.org/lite/models/modify/model_maker/text_searcher
  - https://github.com/google-research/google-research/tree/master/scann
- on-device machine learning
  - [Edge-Adaptable Serverless Acceleration for Machine Learning IoT Applications](https://sites.cs.ucsb.edu/~ckrintz/papers/ZhangSTOIC_SPE20.pdf)
  - [Advantages and limitations of on-device ML](https://developers.google.com/learn/topics/on-device-ml/learn-more)
- Mobile Wifi and 5G limitations for on-device learning
- literature survey **examples** from prior students
  - [Survey on social reputation mechanisms: Someone told me I can trust you](https://arxiv.org/abs/2212.06436)
  - [Survey of stable coins](https://github.com/rwblokzijl/stablecoin-survey/raw/master/Survey.pdf) {52 citations}
  - [The Universal Trust Machine: A survey on the Web3 path towards enabling long term digital cooperation through decentralised trust](https://arxiv.org/abs/2301.06938)
- Literature survey has always 1 essential taxonomy table. Taxonomy in a single glance. Systematic overview of past years of innovation. Key milestones identified. Scientific grounding, 1 or more scientific article per entry/line/milestone. Table with overview and literature, see brilliant With Honours example: ![image](https://user-images.githubusercontent.com/325224/214526306-8f85d788-08e0-45da-9695-f92e33f78f98.png)

Very concrete _possible_ thesis direction. Starting point: take the proof-of-principle implementation of a Web3 recommender with full decentralisation and monte-carlo basic machine learning. Replace the recommender with TFLite model, add hardware-acceleration, tune the memory usage toward huge trust graphs, tune algorithm to giant content databases,  and make it "industrial-proof". Actual deployment to Google Play store and watch the crash reports.  Starting point: "[Web3Recommend: Decentralised recommendations with trust and relevance](https://arxiv.org/pdf/2307.01411)" by Delft student Rohan. His repo: https://github.com/rmadhwal/trustchain-superapp/tree/TrustedRecommendations

_update_: https://www.economist.com/leaders/2023/05/11/what-does-a-leaked-google-memo-reveal-about-the-future-of-ai
_update2_: [BeyondFederated by master student Quinten:](https://github.com/Tribler/tribler/issues/7254#issuecomment-1623096757)
_update3_: First month of thesis time is wisely spend _actively_ {e.g. coding&reading} exporing 2 diffirent thesis direction and may end up inspired for your 3rd direction. {Decentralised federated: hardware-acceleration, tune stochastic gradient decent for infinite scalability, focus specifically on decentralised content recommendation, or learning rate and parameter tuning magic}
Update4: 2011 earliest work. [Gossip Learning with Linear Models on Fully Distributed Data](https://arxiv.org/abs/1109.1396)

**Comment by AkiSchmatzler** [comment]
[book on advances of federated learning](https://arxiv.org/pdf/1912.04977.pdf)

**Comment by synctext** [comment]
Ambition level is more phd thesis instead of master thesis :game_die: 
Quick search on existing code:
- as starting point: get TensorFlow going on _your_ Raspberry Pi:
  - https://github.com/tertiarycourses/RaspberryPITensorflow
  - https://github.com/smirko-dev/machine-learning-rpi
  - https://github.com/Qengineering/RPi-image
